datafusion/core/src/datasource/listing/table.rs

```diff
@@ -220,6 +225,16 @@ pub struct ListingOptions {
     /// Group files to avoid that the number of partitions exceeds
     /// this limit
     pub target_partitions: usize,
+    /// Optional pre-known sort order. Must be `SortExpr`s.
+    ///
+    /// DataFusion may take advantage of this ordering to omit sorts
+    /// or use more efficient algorithms. Currently sortedness must be
+    /// provided if it is known by some external mechanism, but may in
+    /// the future be automatically determined, for example using
+    /// parquet metadata.
+    ///
+    /// See <https://github.com/apache/arrow-datafusion/issues/4177>
+    pub file_sort_order: Option<Vec<Expr>>,
```

Comment:
As I was working on this feature, it occurs to me that if the sort order is to be maintained, the ListingTable can't put multiple file into the same partition 🤔  


The ListingTable does not attempt to read all files concurrently instead it will read files in sequence within a partition.  This is an important property as it allows plans to run against 1000s of files and not try to open them all concurrently.

However, it means if we assign more than one file to a partitition the output sort order will not be preserved as illustrated in this diagrams

When only 1 file is assigned to each partition, each partition is correctly sorted on `(A, B, C)`

```text
┏ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ┓
  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐ ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ┐
┃   ┌───────────────┐     ┌──────────────┐ │   ┌──────────────┐ │   ┌─────────────┐   ┃
  │ │   1.parquet   │ │ │ │  2.parquet   │   │ │  3.parquet   │   │ │  4.parquet  │ │
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │   │Sort: A, B, C │ │   │Sort: A, B, C│   ┃
  │ └───────────────┘ │ │ └──────────────┘   │ └──────────────┘   │ └─────────────┘ │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘  ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘  ─ ─ ─ ─ ─ ─ ─ ─ ─  ┃
     DataFusion           DataFusion           DataFusion           DataFusion
┃    Partition 1          Partition 2          Partition 3          Partition 4       ┃
 ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━

                                      ParquetExec
```


When more than 1 file is assigned to each partition, each partition is NOT correctly sorted on `(A, B, C)`. Once the second file is scanned, the same values for A, B and C can be repeated in the same sorted stream

```text
┏ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━
  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐ ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┃
┃   ┌───────────────┐     ┌──────────────┐ │
  │ │   1.parquet   │ │ │ │  2.parquet   │   ┃
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │
  │ └───────────────┘ │ │ └──────────────┘   ┃
┃   ┌───────────────┐     ┌──────────────┐ │
  │ │   3.parquet   │ │ │ │  4.parquet   │   ┃
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │
  │ └───────────────┘ │ │ └──────────────┘   ┃
┃                                          │
  │                   │ │                    ┃
┃  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
     DataFusion           DataFusion         ┃
┃    Partition 1          Partition 1
 ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ┛

              ParquetExec
```


Does anyone have a preference between any of the possibilities:

1. Adding a check in DataFusion so that sorts will be ignored if there are more than one file in each partition? 
2. Making more partitions if a sort order is specified
3. Making more partitions up to a point (maybe 2x the target partitions and if not ignore the sort order)?

This affects IOx, see https://github.com/influxdata/influxdb_iox/issues/6125

Reply:
🤔  after thinking about this over the evening I think I like option 1 (ignore provided sort order if there is more than one file in each partition) the best

Reply:
I would prefer option (1) as well. 
In future, we can implement a new SortPreservingParquetExec which can read multiple pre-sorted parquet files in a partition and leverage the efficient merge sort to keep the sort ordering. And during physical planning time, if the parent plan do needs the sort info, the physical planner can choose the SortPreservingParquetExec instead of normal ParquetExec.

Reply:
BTW, I don't see the logic to ignore the sort info if there are more than one files in each partition in this PR. 


````
let (partitioned_file_lists, statistics) =
            self.list_files_for_scan(ctx, filters, limit).await?;

        // if no files need to be read, return an `EmptyExec`
        if partitioned_file_lists.is_empty() {
            let schema = self.schema();
            let projected_schema = project_schema(&schema, projection.as_ref())?;
            return Ok(Arc::new(EmptyExec::new(false, projected_schema)));
        }
````

Reply:
You are correct -- I will add the logic to ignore sort order to this PR

Reply:
Added